### ISSUE TITLE
OpenSCAD has switched to French for macOS Sierra

### ISSUE THREAD

**Issue by mprogers** [openscad/openscad#1813]
After upgrading to macOS Sierra, OpenSCAD's menus are all in French, even though my primary language is set to English and my region to the US. Avez-vous des idées? (Do you have any ideas?)

TODO:
- [x] Reproduce on dev machine
- [x] Validate that the issue is related to env. variables

BountySource: https://www.bountysource.com/issues/38026205-openscad-has-switched-to-french-for-macos-sierra

**Comment by t-paul** [comment]
Can you please open a Terminal window, run `locale` and paste the output of that here?

**Comment by mprogers** [comment]
> On Oct 1, 2016, at 11:17 AM, Torsten Paul <notifications@github.com> wrote:
> 
> Can you please open a Terminal window, run locale and paste the output of that here?
> 
> 
Yes, here you go:

$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=

**Comment by t-paul** [comment]
That's strange. What happens when running OpenSCAD from a terminal? Is that still giving the translated texts? e.g. assuming it's in the user Applications folder:

`$HOME/Applications/OpenSCAD.app/Contents/MacOS/OpenSCAD`

As workaround it's possible to disable translations in the preferences completely which should give the default English text again.

**Comment by mprogers** [comment]
> On Oct 1, 2016, at 11:36 AM, Torsten Paul <notifications@github.com> wrote:
> 
> That's strange. What happens when running OpenSCAD from a terminal? Is that still giving the translated texts?
> 
No, that opens up in English.

Michael

**Comment by Pointedstick** [comment]
I have the same issue, and opening from Terminal loads it in English.

```
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

**Comment by kintel** [comment]
Does anyone know how to install macOS Sierra on an OS X VirtualBox host? (i.e. not someone who can Google it, but someone who actually did it)

**Comment by doug-moen** [comment]
https://developer.apple.com/library/content/technotes/tn2418/_index.html

"With iOS 9, the results returned by Locale.preferredLanguages() can differ
from previous releases. In iOS 8 and earlier, only certain language,
script, and region combinations were returned by this API. However, in iOS
9, more combinations of language, script, and region are permitted. This
change is also applied to macOS Sierra, watchOS 3, and tvOS 10.

For example, when a user has configured their iOS device with language set
to English and region set to India, Locale.preferredLanguages() will now
return [ "en-IN" ], instead of [ "en" ]. This allows for smarter language
fallbacks; for this user, if an app doesn’t support en-IN as a
localization, but does support en-GB, the fallback mechanism will select
en-GB instead of en."

**Comment by kintel** [comment]
We use gettext and setlocale() to manage translations. I don't think any of those would call into the Locale library. ..unless Apple simply didn't document any changes to libc..

**Comment by ipartola** [comment]
So... has there been a fix or a workaround for this?

**Comment by t-paul** [comment]
Fix, no, it's still not clear what change causes this. Workarounds only the 2 mentioned above.

**Comment by kintel** [comment]
@ipartola I would look into it if I could get access to a dev machine running Sierra. ..or if someone could help getting Sierra to install in VirtualBox on a Yosemite host.

**Comment by donbright** [comment]
Here is a brief list of some other programs facing similar issues.

**Xamarin**

https://forums.xamarin.com/discussion/76678/xs-changed-language-after-upgrade-to-macos-sierra-how-to-fix

According to Marius Ungureanu on the linked site,,"this is a bug in OS X Sierra caused by regressions in OS X's internationalization stack. We have deployed a fix on our side for the alpha channel"

**Matlab**

(warning, harsh thread.)
https://www.mathworks.com/matlabcentral/answers/303757-why-does-matlab-fail-to-load-my-preferred-language-on-macos-sierra?requestedDomain=www.mathworks.com

"Apple released macOS Sierra on September 20, 2016 and changed the way macOS reports Language & Region preferences to MATLAB."

note that the Matlab people are linking to the same Technical Note 2418 page as @doug-moen did above. https://developer.apple.com/library/content/technotes/tn2418/_index.html

**Lilypond**

http://lilypond.1069038.n5.nabble.com/language-issues-with-Mac-OS-X-Sierra-10-12-td195168.html

Doesn't look like they have a complete fix yet. Note Lilypond is a GNU / open source project so probably the most similar to OpenSCAD in terms of language backend as compared to say closed source projects like MatLab. 

***iOS9 apps***

If you type into google "iOS9 localization bug" you will find a huge amount of hits, many of them link to that same Technical Note 2418 https://developer.apple.com/library/content/technotes/tn2418/_index.html

Example

http://stackoverflow.com/questions/32710913/ios-localization-is-broken-with-the-upgrade-ios-9-xcode-7/33524891

**Comment by t-paul** [comment]
Thanks for digging, that's really some nasty discussions going on, especially in that Matlab thread.

I'm still not sure this is exactly the same issue we see. The Matlab thread also mentions:
```
The following locales are not impacted by this issue:

    * English, United States
    ...
```
But that's exactly the locale shown on command line in the cases listed above as `en_US.UTF-8`. @mprogers, @Pointedstick could you please check if the configuration is indeed one of the locales listed in the [Answer in the Matlab forum](https://de.mathworks.com/matlabcentral/answers/303757-why-does-matlab-fail-to-load-my-preferred-language-on-macos-sierra?#answer_235386) (e.g. showing "English, United States" as the top-most entry)?

I guess one way to find out is to add the locale values to our library info screen and see what that shows when the application is started from finder.

**Comment by Pointedstick** [comment]
My top locale appears to be English - U.S. 
![screen shot 2016-10-21 at 7 18 47 am](https://cloud.githubusercontent.com/assets/1097249/19599579/a7dc46ca-975e-11e6-9eb1-07eb3784d427.png)


I continue to see the issue as long as I have localization checked in OpenSCAD's preferences.

**Comment by kintel** [comment]
@Pointedstick What happens if you remove all languages except English from your preferred languages?

**Comment by Pointedstick** [comment]
@kintel Aha! removing all but English makes OpenScad launch properly in English even with localization is turned on.

**Comment by kintel** [comment]
@Pointedstick Thanks for testing. Yes, the problem only occurs when launching from the Finder.

**Comment by bbum** [comment]
When launched from the Finder, there really isn't a standard shell like environment.   In past OS X releases, more of the user's shell configuration was made available to apps that were double-clicked in the Finder.   That has changed over time and it appears that the locale was the latest to be dropped.

You'll need a bit of OS X specific glue to grab the localizations from NSLocale.

**Comment by kintel** [comment]
@bbum That would indeed explain a lot.

**Comment by kintel** [comment]
This turned out to be a bug in gettext (https://savannah.gnu.org/bugs/?49560). Awaiting an update to gettext, this has been fixed by patching our gettext build.
Note: This means that until other builds of gettext are patched (e.g. homebrew, MacPorts), only the official OS X binaries of OpenSCAD will work as expected.

**Comment by MichaelAtOz** [comment]
@t-paul @kintel 
So who should claim the bounty for this?

**Comment by kintel** [comment]
Can we refund it? It was fixed by someone in the `gettext` library..

**Comment by MichaelAtOz** [comment]
There doesn't seem to be anything built-in to do a refund. I could email BS.

_Or_, for the effort ($9), @donbright @t-paul & @kintel seemed to share the co-ord for this;  **$3 each? Buy yourselves  [something special](https://www.shapeways.com/product/6VEMXWHGV/logo2-15mm?key=e260d6e6f5bd261cda619158b02b4064&optionId=65365611)** :1st_place_medal:  -  (the metalic plastic is < $3) - at cost ATM for those who see this.

The 18k Gold Plated polishes up nice (this is the 10mm long version):
![logo2 10 gold walnut](https://user-images.githubusercontent.com/1972961/39415166-9013e49e-4c2f-11e8-8041-3cede6cd5840.JPG)

Or you may want to save up for the developer moonlight ninja edition ( 15mm raw bronze): 
![moonlight raw bronze](https://user-images.githubusercontent.com/1972961/39415545-1f3080ca-4c34-11e8-8aa8-351d4508516d.JPG)
(please excuse the phone camera limitations)

**Comment by MichaelAtOz** [comment]
Or, I just found this in the FAQs
```
If the issue was closed as a "won't fix", or is deemed not in line with the project's goals, 
the bounty is refunded.

If I posted a bounty using team funds, where does the refund go?
Refunds will go back into the team account.
```

Could try reopening, labelling Won't fix & closing.

**Comment by t-paul** [comment]
Refund sounds good to me. As far as I can see the bounty was posted by @kintel anyway, so just claiming it would be same effect, I guess?

**Comment by ErnestoBorio** [comment]
As you can see in the picture, my preferred language is english and locale is en-US.
Only after removing all preferred languages and leaving only english did it.
Before this, Russian was chosen, which was like 5th on the preferred language list.
Also, nothing like this has ever happened to me in 10 years of using MacOS, going back to 10.5 Leopard, so I guess it's not OS standard.

![screen shot 2018-07-16 at 03 01 26](https://user-images.githubusercontent.com/1771719/42747593-79c9c30a-88b3-11e8-98a7-55b5b0500ab5.png)

**Comment by kintel** [comment]
@ErnestoBorio This was a bug in `gettext` (a 3rd party library). It's fixed in our latest development snapshots.

**Comment by ErnestoBorio** [comment]
Something I didn't know is that when you add language inputs on MacOS, the OS adds those languages to the preferred language list without even telling you.
I think this may go unadvertedly for most users as usually when you are studying a new language you like to add its keyboard input maybe just for fun or studying but you hardly know the language, and you certainly do not want software UI to be in it.

**Comment by Diti** [comment]
I’m on Windows with an English locale, and OpenSCAD is set to French for some reason. Does anybody know how to switch it to English?

**Comment by kintel** [comment]
Please open a new issue. This was about macOS and turned out to be a Mac-specific bug in 3rd party software.
